### PR TITLE
HASS Plugin Power management support

### PR DESCRIPTION
**Pull request by stickpin**
Power management is disabled by default.
To enable power management you should check the "Allow to Home Assistant to use power management functionality" checkbox in the Integration Options.

Screenshot 2020-03-20 at 11 14 13

@postlund please review the code and let me know if something is missing or you want me to change.
Thanks in advance!

### CODE REVIEW

File: custom_components/apple_tv/__init__.py [hunk 1]
```diff
@@ -151,6 +153,8 @@ def __init__(self, hass, config_entry):
         self.message = None
         self.atv = None
         self._is_on = not config_entry.options.get(CONF_START_OFF, False)
+        self._is_pwr_mgmt_on = config_entry.options.get(CONF_PWR_MGMT, False)
+        self._is_already_connected = False
```
Comment (postlund) [inline custom_components/apple_tv/__init__.py]: You can just check if `self.atv` is set instead of this.

File: custom_components/apple_tv/__init__.py [hunk 2]
```diff
@@ -151,6 +153,8 @@ def __init__(self, hass, config_entry):
         self.message = None
         self.atv = None
         self._is_on = not config_entry.options.get(CONF_START_OFF, False)
+        self._is_pwr_mgmt_on = config_entry.options.get(CONF_PWR_MGMT, False)
```
Comment (postlund) [inline custom_components/apple_tv/__init__.py]: This is a bit tricky. Doing it like this force the user to restart Home Assistant in order for the option to take effect. One way is to read this value from `self.config_entry` every time. But since config entries supports listeners, we should listen for updates and update variable accordingly.

https://github.com/home-assistant/core/blob/dev/homeassistant/config_entries.py#L381

File: custom_components/apple_tv/__init__.py [hunk 3]
```diff
@@ -177,8 +181,19 @@ def connection_closed(self):
 
     async def connect(self):
         """Connect to device."""
-        self._is_on = True
-        self._start_connect_loop()
+        if self._is_pwr_mgmt_on and self._is_already_connected:
```
Comment (postlund) [inline custom_components/apple_tv/__init__.py]:
```suggestion
        if self._is_pwr_mgmt_on and self.atv:
```

File: custom_components/apple_tv/__init__.py [hunk 4]
```diff
@@ -177,8 +181,19 @@ def connection_closed(self):
...
+            self._is_on = True
+            self._start_connect_loop()
+
+    async def softdisconnect(self):
+        """Turn off device."""
+        _LOGGER.debug("Turning the device off")
+        await self.atv.power.turn_off()
+        self._update_state(disconnected=True)
 
     async def disconnect(self):
```
Comment (postlund) [inline custom_components/apple_tv/__init__.py]: I would prefer if we could let the manager deal with everything related to power management. So, I suggest that we rename current `disconnect` to `close_connection` and `softdisconnect` to just `disconnect`. Then we add the logic from `media_player` and `remote` in there:

```python
async def disconnect(self):
    """Disconnect from device."""
    if self._is_pwr_mgmt_on:
        _LOGGER.debug("Turning the device off")
        await self.atv.power.turn_off()
        self._update_state(disconnected=True)
    else:
        await self._close_connection()
```
More or less. We can then revert the changes in said platforms.

File: custom_components/apple_tv/__init__.py [hunk 5]
```diff
@@ -307,6 +322,14 @@ def _auth_problem(self):
 
         self.address_updated(str(conf.address))
 
+        if self._is_pwr_mgmt_on:
+            self._is_already_connected = True
+            self.atv.power.listener = PowerListener(self)
+            if self.atv.power.power_state == PowerState.On or self.atv.power.power_state == PowerState.Unknown:
```
Comment (postlund) [inline custom_components/apple_tv/__init__.py]:
```suggestion
            if self.atv.power.power_state in [PowerState.On, PowerState.Unknown]:
```

File: custom_components/apple_tv/__init__.py [hunk 6]
```diff
@@ -339,3 +362,19 @@ def address_updated(self, address):
...
+class PowerListener:
+    """Listener interface for power updates."""
+
+    def __init__(self, manager):
+        """Initialize the Apple TV device."""
+        self.atv = None
+        self._manager = manager
+
+    def powerstate_update(self, old_state, new_state):
+        if new_state == PowerState.On or new_state == PowerState.Unknown:
```
Comment (postlund) [inline custom_components/apple_tv/__init__.py]:
```suggestion
        if new_state in [PowerState.On, PowerState.Unknown]:
```

File: custom_components/apple_tv/media_player.py [hunk 7]
```diff
@@ -215,7 +215,10 @@ def supported_features(self):
     async def async_turn_off(self):
         """Turn the media player off."""
         self._playing = None
-        await self._manager.disconnect()
+        if self._manager._is_pwr_mgmt_on:
```
Comment (postlund) [inline custom_components/apple_tv/media_player.py]: Revert according to above.

File: custom_components/apple_tv/remote.py [hunk 8]
```diff
@@ -83,7 +83,10 @@ def should_poll(self):
 
     async def async_turn_off(self, **kwargs):
         """Turn the device off."""
-        await self._manager.disconnect()
+        if self._manager._is_pwr_mgmt_on:
```
Comment (postlund) [inline custom_components/apple_tv/remote.py]: Revert according to above.

**Comment by stickpin** [comment]
@postlund working to apply all the fixes now.
I am not so sure how I've caused the conflict and how I can rebase the code. Sorry I am still kinda new to all this git stuff.

**Comment by stickpin** [comment]
@postlund seems like the same way I've created conflict, the same way I've resolved it. :)
After the last commit, it's no longer present.
Two things left, config part and tests. Once you will let me know that the main code is ok, I will start working on tests. :)

**Comment by postlund** [comment]
It is probably because you merge back changes from the origin branch. It works as well but I will have to create a squashed commit before merging it back to master. It's a matter of taste.

My personal preference is to keep a straight history and avoid merges altogether. It tends to make it a lot easier to keep track of your own and other people changes. You do that by simply never merge stuff. When you want to bring in updates, you rebase on top of the origin branch, e.g. master.

To start a new feature branch:
* git co master
* git pull # make sure master is up-to-date
* git co -b my_feature

Now  you are on `my_feature` branch. Make your changes and commit. When you want to bring in updates that has been committed to master:

* git checkout master
* git pull
* git checkout my_feature
* git rebase master

Git will remove all commits from your branch, change to the latest commit on master and re-apply all commits again. If you get a conflict, solve it, do "git add" on those files and then "git rebase --continue". I would say that this is much better (and more commonly used flow).

Short-form to create feature branch:
* git fetch && git co -b feature origin/master

To make it up-to-date:
* git fetch && git rebase origin/master

You will have to force-push (git push -f) since you have re-written the history. Worth remembering.

**Comment by postlund** [comment]
@stickpin It's because I enabled squashed commits😊 You can probably start working on the tests, I won't have much time reviewing this weekend. Could you also do some testing with device reboots and other odd-balls, just to make it sure behaves correctly?

**Comment by stickpin** [comment]
@postlund sure, I will play with different edge cases after I will figure out how to do a config part.
I will keep you posted. 
Keep safe and away from crowd places! :)

**Comment by postlund** [comment]
@stickpin Sounds good 😊 Should be more or less the same as the last test case in the test file.

Same too you! You can never be too careful!

File: custom_components/apple_tv/__init__.py [hunk 9]
```diff
@@ -364,6 +365,12 @@ def address_updated(self, address):
         )
         self.hass.add_job(update_entry, self.config_entry)
 
+    @staticmethod
+    async def async_options_updated(hass, config_entry):
```
Comment (stickpin) [inline custom_components/apple_tv/__init__.py]: @postlund can you assist, please? :)
I've added config_entry update listener and it's working as expected.
The only problem that I am not able to pass `self` to `async_options_updated` so I will be able to change `self._is_pwr_mgmt_on`.

Thanks in advance.

Comment (postlund) [reply custom_components/apple_tv/__init__.py]: Try removing @staticmethod an add self as first argument.

Comment (stickpin) [reply custom_components/apple_tv/__init__.py]: getting following error:
```
  File "/usr/local/lib/python3.7/site-packages/aiohttp/web_protocol.py", line 418, in start,
    resp = await handler(request),
  File "/usr/local/lib/python3.7/site-packages/aiohttp/web_app.py", line 458, in _handle,
    resp = await task,
  File "/usr/local/lib/python3.7/site-packages/aiohttp/web_middlewares.py", line 119, in impl,
    return await handler(request),
  File "/usr/src/homeassistant/homeassistant/components/http/real_ip.py", line 39, in real_ip_middleware,
  File "/usr/src/homeassistant/homeassistant/components/http/ban.py", line 72, in ban_middleware,
  File "/usr/src/homeassistant/homeassistant/components/http/auth.py", line 127, in auth_middleware,
  File "/usr/src/homeassistant/homeassistant/components/config/config_entries.py", line 227, in post,
  File "/usr/src/homeassistant/homeassistant/components/http/view.py", line 123, in handle,
    result = await result,
    return await super().post(request, flow_id),
  File "/usr/src/homeassistant/homeassistant/components/http/data_validator.py", line 50, in wrapper,
  File "/usr/src/homeassistant/homeassistant/helpers/data_entry_flow.py", line 92, in post,
    result = await method(view, request, *args, **kwargs),
    result = await self._flow_mgr.async_configure(flow_id, data),
  File "/usr/src/homeassistant/homeassistant/data_entry_flow.py", line 130, in async_configure,
  File "/usr/src/homeassistant/homeassistant/data_entry_flow.py", line 198, in _async_handle_step,
    result = await self._async_handle_step(flow, cur_step["step_id"], user_input),
    result = await self.async_finish_flow(flow, dict(result)),
  File "/usr/src/homeassistant/homeassistant/config_entries.py", line 935, in async_finish_flow,
    self.hass.config_entries.async_update_entry(entry, options=result["data"]),
  File "/usr/src/homeassistant/homeassistant/config_entries.py", line 747, in async_update_entry,
    self.hass.async_create_task(listener(self.hass, entry)),
TypeError: 'NoneType' object is not callable,
2020-03-21 19:17:14 ERROR (MainThread) [aiohttp.server] Error handling request,
Traceback (most recent call last):
```

Comment (postlund) [reply custom_components/apple_tv/__init__.py]: Hmm, that makes it a bit more "complicated". Let me look into it and I'll get back to you.

Comment (postlund) [reply custom_components/apple_tv/__init__.py]: Ok, so, I believe we should do it like this. Store the value as for instance `self.is_pwr_mgmt_on` in the manager class. Then write a callback method similar to this:

```python
    @staticmethod
    async def async_options_updated(hass, entry):
        """Triggered by config entry options updates."""
        hass.data[DOMAIN][entry.entry_id].is_pwr_mgmt_on = entry.options[CONF_PWR_MGMT]
```

Should work.

Comment (stickpin) [reply custom_components/apple_tv/__init__.py]: I've already tried a similar solution and I've got the following error:
```
2020-03-23 10:27:31 ERROR (MainThread) [homeassistant.core] Error doing job: Task exception was never retrieved,
Traceback (most recent call last):,
    hass.data[DOMAIN][entry.entry_id].is_pwr_mgmt_on = entry.options[CONF_PWR_MGMT],
  File "/config/custom_components/apple_tv/__init__.py", line 371, in async_options_updated,
KeyError: '2dc26103d96a4b37801a1221a0bc7ed0'
```

Comment (postlund) [reply custom_components/apple_tv/__init__.py]: Ah, right. It should be `entry.unique_id` and not `entry.entry_id`.

Comment (stickpin) [reply custom_components/apple_tv/__init__.py]: @postlund yep, that one seems to work.
I will test it and push the fix.

**Comment by stickpin** [comment]
@postlund now it should be good.
I've also covered the cases when Apple TV was On or Off during the options change, the same goes for HASS Remote or Media instance.
Please let me know if it looks good, so I will start working on the tests.

File: custom_components/apple_tv/__init__.py [hunk 10]
```diff
@@ -339,3 +365,31 @@ def address_updated(self, address):
             data={**self.config_entry.data, CONF_ADDRESS: address},
         )
         self.hass.add_job(update_entry, self.config_entry)
+
+    @staticmethod
+    async def async_options_updated(hass, entry):
+        """Triggered by config entry options updates."""
+        hass.data[DOMAIN][entry.unique_id]._is_pwr_mgmt_on = entry.options[CONF_PWR_MGMT]
+        if entry.options[CONF_PWR_MGMT] and hass.data[DOMAIN][entry.unique_id].atv:
```
Comment (postlund) [inline custom_components/apple_tv/__init__.py]: I'm not sure about this functionality (from here and until end of function), is there any particular reason for doing this?

**Comment by postlund** [comment]
@stickpin Just a small question on my end, otherwise it looks good to me!

File: custom_components/apple_tv/__init__.py (hunk 10 above)
Comment (stickpin) [reply custom_components/apple_tv/__init__.py]: @postlund it's covering the following situations...
For example:
Apple TV power was On and HASS Remote/Media Control switch was Off (`atv = None`). If you changing Option to `Allow to Home Assistant to use power management functionality` this code will create `atv` instance and will update the `power_state` to the real one.

The same goes for the vice-versa scenario.

I hope it's clear.

Comment (stickpin) [reply custom_components/apple_tv/__init__.py]: @postlund all good or you have more questions? 
can I progress with the tests? :)

Comment (postlund) [reply custom_components/apple_tv/__init__.py]: I'm still not a 100% sure about having the callback trigger power on/off automatically like but. But I'll leave it for the Home Assistant devs that does the review later on. We can say that it's ok! Carry on with the tests!

**Comment by postlund** [comment]
Oh, we need to reflect power state back to Home Assistant somehow. Currently, if the device is "off", it will be reported as "idle" and that is not very helpful to user wanting to write automations based on power state. Can you fix that?

**Comment by stickpin** [comment]
@postlund, sorry mate, I currently have no time to work on it. 
It's a quite challenging time now at work, hopefully it will over soon. If you have no time to complete this PR, let's keep it on hold for now.

**Comment by postlund** [comment]
@stickpin No worries, no rush here as the PR in Home Assistant isn't even merged yet 😄 Lets continue when there's more time!

**Comment by cobch** [comment]
@stickpin Any news about the power management of ATV via hass? It will be so great to have it in order to be able to make more automations (-:

**Comment by stickpin** [comment]
@cobch I will try to restart my work on it this week.
For time being you can install and use it from my repo: https://github.com/stickpin/hass-atv-beta

@postlund Hi. I hope you well there. :) How we want to restart the process? :)

**Comment by postlund** [comment]
@stickpin Yeah, I'm fine 😊 I suggest fixing the stuff I mentioned before (think it was something about state). Then it would be good to get some external users to verify that it works before merging the PR.

**Comment by stickpin** [comment]
@postlund, any chance you can get me up to speed?

I am not sure what has been changed since I was working on the power management implementation on the plugin side, but it seems that with the latest `pyatv` PowerListener interface for some reason no longer working at least the way I've implemented it.

**Comment by postlund** [comment]
@stickpin Yeah, I think it's the change to weak references to listeners. Have a look at https://pyatv.dev/support/troubleshooting/ and scroll down to "I do not receive updates...".

**Comment by stickpin** [comment]
@postlund maybe I am missing something but no matter which combination of code I try, the listener is not triggered.

**Comment by postlund** [comment]
@stickpin You need to make sure someone has a reference to the `PowerListener` instance after assigning it to the listener property, otherwise it will be garbage collected. You can create it in the constructor and keep a reference to it from there for instance.

**Comment by stickpin** [comment]
@postlund, and this is what I did in the _connect function:
```
if self._is_pwr_mgmt_on:
            power_listener = PowerListener(self)
            self.atv.power.listener = power_listener
            if self.atv.power.power_state in [PowerState.On, PowerState.Unknown]:
                self._update_state(connected=True)
            else:
                self._update_state(disconnected=True)
```

I see that It's getting to `__init__` but `powerstate_update` is not triggered.

**Comment by postlund** [comment]
You need to do:
```python
self.power_listener = PowerListener(self)
```
In the constructor or no one will have a reference to it once it goes out of scope.

**Comment by stickpin** [comment]
now I am completely lost... :) 
This is what I had from the beginning:
`self.atv.power.listener = PowerListener(self)`

what is the idea to do it in the following way?:
```
self.power_listener = PowerListener(self)
self.atv.power.listener = self.power_listener
```

sorry for bothering... just want to understand the logic behind it...

**Comment by postlund** [comment]
@stickpin When you assign an object to a variable, that object will get a reference count of one (or +1 to be exact). This is what happens when you do `self.power_listener = ...`. This will make sure that the object isn't garbage collected, since that would be very bad. Once you do `self.power_listener = None` (explicit in this case, but could be implicit too), the reference count will be decreased and reach zero. Meaning that no one can "reach" the object and it will/can be garbage collected. This is the basics of all memory management in python and other garbage collected languages. 

What I use here is a weak reference. It does not alter the reference count at all. So if you store a "weak reference" to an object and that object goes out of scope, then the object can be garbage collected (since the reference count will reach zero) and the weak reference will not point to anything. A normal use case for this is object caching. An object (usually quite large) can be stored in a cache using a weak reference, meaning that it can be available in the cache but the garbage collector can also free it at any time.

I made this change so that users doesn't have to unassign listeners. Failing to do so could make pyatv keep objects around that aren't really used anymore. So the approach is more that you can assign listeners and if they are still alive when something happens, then they will be called. Otherwise the call is ignored.

**Comment by stickpin** [comment]
@postlund I think the last commit should be stable enough to start testing.
Please let me know if you something strange there.

**Comment by postlund** [comment]
@stickpin I will have a look later tonight! 👍

File: custom_components/apple_tv/__init__.py [hunk 11]
```diff
@@ -339,3 +400,31 @@ def address_updated(self, address):
             data={**self.config_entry.data, CONF_ADDRESS: address},
         )
         self.hass.add_job(update_entry, self.config_entry)
+
+    @staticmethod
+    async def async_options_updated(hass, entry):
+        """Triggered by config entry options updates."""
+        hass.data[DOMAIN][entry.unique_id]._is_pwr_mgmt_on = entry.options[CONF_PWR_MGMT]
```
[the remark this answers is not in the document]
Comment (stickpin) [reply custom_components/apple_tv/__init__.py:407]: Never did squash before but will try to do it now.
Regarding `hass.data[DOMAIN][entry.unique_id]` will change it after squash.

Regarding tests, how execute them?

[the remark this answers is not in the document]
Comment (postlund) [reply custom_components/apple_tv/__init__.py:407]: I use the run.sh (which is not ideal and a bit out-of-dats) script to run tests. It requires that you have homeassistant checked out next to hass-atv-beta. You can modify it and remove the "../conf" volume as it is not needed. You don't have to provide any additional arguments, just run the script.

**Comment by stickpin** [comment]
@postlund closing this PR, re-submitted a new [PR#5](https://github.com/postlund/hass-atv-beta/pull/2) .
Re-forking and updating the code from scratch was faster than to learn how to do squash. :)

**Comment by postlund** [comment]
@stickpin Yeah, at this stage it would probably have been easier to just check out master and put all the changes on top of it (copy file content from this branch) and force pushed.
